README.md
# todotxt-mode README

"todotxt-mode" is an extension for working with [todo.txt](https://github.com/todotxt/todo.txt) formatted files. It includes formatting of the different task "axes": *priority*, *project* and *context*, as well as a number of commands for automating and simplifying useful tasks and common workflows.

This extension was inspired by the [todo-txt](https://marketplace.visualstudio.com/items?itemName=rarnoldmobile.todo-txt) VS Code extension and borrows some ideas and features from it, but takes it in a direction to make it easier to implement a GTD-like process using `todo.txt`-formatted files.

## Features

### Highlighting
The extension uses decorators to highlight tags of task items.

- Syntax highlighting of task priority (`(A)`), creation date, context (`@label`) and project (`+label`) tags
- Priority format can vary among A, B and C-Z levels
- Syntax higlighting of metadata tags (`key:value1`)
- Additional formatting for `due:YYYY-MM-DD` date for past, present and future dates

![Highligted Tasks](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/highlightedtasks.png)


### Task Management
Commands that help manage the task list include
- Mark tasks as complete
- Sort tasks by project, priority, context, creation date, due date or metadata tag
- Stable sort so tasks without the sort key are not affected
- Reformat task tags to a consistent sequence

![Manage Tsks](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/sort-complete.gif)

### File Workflow Commands
There are a number of commands in the extension to support a GTD-like workflow.

- Move tasks from any file to the main *Todo* file (used when processing the *Inbox* file)
- Move tasks to the *Waiting* file
- Move tasks to the *Someday* file
- Move completed tasks to the *Archive* file

![Manage Tsks](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/move-tasks.gif)

### Todo.txt enhancements
The extension also includes a few enhancements that, while not part of the todo.txt spec, can be useful.
- Optional delimiter to break file into tasks section and notes section
- Save selection to a note and add to a task as a `note:<file>` metadata tag, with preview

![Create Note](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/create-note.gif)
![Preview Note](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/preview-note.gif)

- Highlight a section of a Markdown file that contains tasks

![Markdown Highlighting](https://raw.githubusercontent.com/davraamides/todotxt-mode/master/images/highlightedmarkdown.png)

## Extension Settings

### Commands

| Command                           | Description                                                       |
| --------------------------------- | ----------------------------------------------------------------- |
| `todotxt-mode.toggleCompletion`   | Toggle a task as complete                                         |
| `todotxt-mode.sortByContext`      | Sort all/selected tasks by context - `@label`                     |
| `todotxt-mode.sortByPriority`     | Sort all/selected tasks by priority - `(A)`                       |
| `todotxt-mode.sortByProject`      | Sort all/selected tasks by project - `+label`                     |
| `todotxt-mode.sortByTag`          | Sort all/selected tasks by metadata tag - `key:value`             |
| `todotxt-mode.sortByCreationDate` | Sort all/selected tasks by creation date                          |
| `todotxt-mode.sortByDueDate`      | Sort all/selected tasks by due date - `due:YYYY-MM-DD`            |
| `todotxt-mode.formatTasks`        | Format tasks with all tags in consistent order at end of the line |
| `todotxt-mode.archiveTasks`       | Move completed tasks to the *Done* file                           |
| `todotxt-mode.moveTasksToTodo`    | Move selected tasks to the *Todo* file                            |
| `todotxt-mode.moveTasksToWaiting` | Move selected tasks to the *Waiting* file                         |
| `todotxt-mode.moveTasksToSomeday` | Move selected tasks to the *Someday* file                         |
| `todotxt-mode.moveTasksToProject` | Move selected tasks to a Markdown project file                    |
| `todotxt-mode.createTaskNote`     | Create a Markdown note file from the selected note text           |
| `todotxt-mode.incrementPriority`  | Increment priority of selected tasks                              |
| `todotxt-mode.decrementPriority`  | Decrement priority of selected tasks                              |
| `todotxt-mode.removePriorities`   | Remove all priorities from tasks                                  |

### Styles

| Setting                                       | Description                                   | Default                                                         |
| --------------------------------------------- | --------------------------------------------- | --------------------------------------------------------------- |
| `todotxtmode.contextStyle.light.color`        | Color of context field in light mode          | <span style="color:rgb(40, 161, 86);">rgb(40, 161, 86)</span>   |
| `todotxtmode.contextStyle.dark.color`         | Color of context field in dark mode           | <span style="color:rgb(40, 161, 86);">rgb(40, 161, 86)</span>   |
| `todotxtmode.highPriorityStyle.light.color`   | Color of high priority field in light mode    | <span style="color:rgb(240, 226, 25);">rgb(240, 226, 25)</span> |
| `todotxtmode.highPriorityStyle.dark.color`    | Color of high priority field in dark mode     | <span style="color:rgb(240, 226, 25);">rgb(240, 226, 25)</span> |
| `todotxtmode.mediumPriorityStyle.light.color` | Color of medium priority field in light mode  | <span style="color:rgb(201, 189, 22);">rgb(201, 189, 22)</span> |
| `todotxtmode.mediumPriorityStyle.dark.color`  | Color of medium priority field in dark mode   | <span style="color:rgb(201, 189, 22);">rgb(201, 189, 22)</span> |
| `todotxtmode.lowPriorityStyle.light.color`    | Color of low priority field in light mode     | <span style="color:rgb(170, 160, 21);">rgb(170, 160, 21)</span> |
| `todotxtmode.lowPriorityStyle.dark.color`     | Color of low priority field in dark mode      | <span style="color:rgb(170, 160, 21);">rgb(170, 160, 21)</span> |
| `todotxtmode.projectStyle.light.color`        | Color of project field in light mode          | <span style="color:rgb(25, 172, 230);">rgb(25, 172, 230)</span> |
| `todotxtmode.projectStyle.dark.color`         | Color of project field in dark mode           | <span style="color:rgb(25, 172, 230);">rgb(25, 172, 230)</span> |
| `todotxtmode.tagStyle.light.color`            | Color of tag field in light mode              | <span style="color:rgb(179, 58, 172);">rgb(179, 58, 172)</span> |
| `todotxtmode.tagStyle.dark.color`             | Color of tag field in dark mode               | <span style="color:rgb(179, 58, 172);">rgb(179, 58, 172)</span> |
| `todotxtmode.creationDateStyle.light.color`   | Color of creation date field in light mode    | <span style="color:rgb(200, 130, 0);">rgb(200, 130, 0)</span>   |
| `todotxtmode.creationDateStyle.dark.color`    | Color of creation date field in light mode    | <span style="color:rgb(200, 130, 0);">rgb(200, 130, 0)</span>   |
| `todotxtmode.pastDateStyle.light.color`       | Color of past due date field in light mode    | <span style="color:rgb(177, 58, 28);">rgb(177, 58, 28)</span>   |
| `todotxtmode.pastDateStyle.dark.color`        | Color of past due date field in dark mode     | <span style="color:rgb(177, 58, 28);">rgb(177, 58, 28)</span>   |
| `todotxtmode.presentDateStyle.light.color`    | Color of present due date field in light mode | <span style="color:rgb(219, 216, 26);">rgb(219, 216, 26)</span> |
| `todotxtmode.presentDateStyle.dark.color`     | Color of present due date field in dark mode  | <span style="color:rgb(219, 216, 26);">rgb(219, 216, 26)</span> |
| `todotxtmode.futureDateStyle.light.color`     | Color of future due date field in light mode  | <span style="color:rgb(118, 194, 31);">rgb(118, 194, 31)</span> |
| `todotxtmode.futureDateStyle.dark.color`      | Color of future due date field in dark mode   | <span style="color:rgb(118, 194, 31);">rgb(118, 194, 31)</span> |
| `todotxtmode.completedStyle.textDecoration`   | Text decoration for completed tasks           | <span style="text-decoration:line-through;">line-through</span> |
| `todotxtmode.completedStyle.opacity`          | Opacity for completed tasks                   | <span style="opacity:0.5;">0.5</span>                           |

### Files

There are a few special files identifed for certain commands

| Setting                       | Description                              | Default       |
| ----------------------------- | ---------------------------------------- | ------------- |
| `todotxtmode.todoFilename`    | Name of main todo text file              | `todo.txt`    |
| `todotxtmode.doneFilename`    | Name of archive file for completed tasks | `done.txt`    |
| `todotxtmode.somedayFilename` | Name of file for someday/maybe tasks     | `someday.txt` |
| `todotxtmode.waitingFilename` | Name of file for waiting tasks           | `waiting.txt` |

### Keybindings

The extension ships with only one keybinding assigned but any of the commands can be bound to keyboard shortcuts in the usual manner with VS Code.

| Command                       | Default Keybinding |
| ----------------------------- | ------------------ |
| `extension.toggleCompletion`  | `ctrl+shift+x`     |
| `extension.incrementPriority` | `ctrl+shift+a`     |
| `extension.decrementPriority` | `ctrl+shift+z`     |


### Other Settings

The remaining settings define various behavior of the extension including which files should have this extension activate.

| Setting                                        | Descripiton                                                                                                                                  | Default                            |
| ---------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------- |
| `todotxtmode.sortCompletedTasksToEnd`          | Alwasy place completed tasks at end of list when sorting                                                                                     | `true`                             |
| `todotxtmode.sectionDelimiterPattern`          | Regex pattern to delimit todo tasks section from other parts of the file                                                                     | `^--$`                             |
| `todotxtmode.todoFilePattern`                  | Regex pattern of filenames to apply todotxt-mode commands                                                                                    | `^.*\\.txt$`                       |
| `todotxtmode.excludeDecorationsFilePattern`    | Regex pattern of filenames to exclude todotxt-mode decorations                                                                               | `^done\\.txt$`                     |
| `todotxtmode.markdownFilePattern`              | Regex pattern of Markdown filenames to apply todotxt-mode commands using `markdownDecorationBeginPattern` and `markdownDecorationEndPattern` | <code>^.*\\.(md\|markdown)$</code> |
| `todotxtmode.markdownDecorationBeginPattern`   | Regex pattern of range in Markdown files to decorate                                                                                         | `^#+\\s+Tasks\\s*$`                |
| `todotxtmode.markdownDecorationEndPattern`     | Regex pattern of range in Markdown files to decorate                                                                                         | `^$`                               |
| `todotxtmode.removePriorityFromCompletedTasks` | Remove the priority tag from completed tasks                                                                                                 | `false`                            |
| `todotxtmode.noteFilenameFormat`               | Format for new note filenames. Can contain strftime-style format specifiers. The string '[Todo]', if used, will be selected.                 | `[Todo]-Note-%Y%m%d-%H%M%S.md`     |
| `todotxtmode.replaceNoteTextWithNoteTag`       | Automatically replace note text with the note tag when creating a note for a task                                                            | `false`                            |
| `todotxtmode.addCreatedAt`                     | Add the creation date when creating a task                                                                                                   | `false`                            |

## Release Notes

See [CHANGELOG](https://github.com/davraamides/todotxt-mode/blob/master/CHANGELOG.md)



CHANGELOG.md
# Change Log

## 1.4.32 - 2023-04-21

- Pull-request ["feat: insert created at on new task"](https://github.com/davraamides/todotxt-mode/pull/55).

## 1.4.30 - 2021-11-16

- Fixed issue on ["Note tag hover shows note content more than once"](https://github.com/davraamides/todotxt-mode/issues/42) where hovers would show note content more than once.

## 1.4.28 - 2021-10-09

- Added `tagDatePattern` setting to allow specifying format of dates in `due` tags. Addresses [Extension Setting to Change Date Format](https://github.com/davraamides/todotxt-mode/issues/33).

## 1.4.27 - 2021-09-28

- Fixed issue on ["Increment/Decrement priority of selected tasks" apply for selected tasks](https://github.com/davraamides/todotxt-mode/issues/30) where the selection was lost after a multi-line edit.

## 1.4.25 - 2021-05-31

- Pull-request to ["Allow projects and contexts to contain non-Latin characters"](https://github.com/davraamides/todotxt-mode/pull/31).

## 1.4.23 - 2021-05-10

- Fixed bug where some multi-task commands like `incrementPriority`, `decrementPriority` and `toggleCompletion` were not working properly. It seems calling `TextEditor.edit` multiple times in a loop no longer works as it used to. This fixes ["Increment/Decrement priority of selected tasks" apply for selected tasks](https://github.com/davraamides/todotxt-mode/issues/30). Updated README.md to include all settings and commands.

## 1.4.21 - 2021-05-07

- Added option to automatically [replace note text with the resulting note tag](https://github.com/davraamides/todotxt-mode/issues/27) when creating a note for a task. The new setting, `replaceNoteTextWithNoteTag`, defaults to `false` so the old behavior remains the same (the text is just deleted).

## 1.4.19 - 2021-02-10

- Added new extension setting to [make the note filename pattern customizable](https://github.com/davraamides/todotxt-mode/issues/25).
    The new settings, `noteFilenameFormat`, specifies the format for new note filenames when using the `createTaskNote` command ("Create note for task"). The format can include `strftime`-style format specifiers and uses the [strftime](https://thdoan.github.io/strftime/) JavaScript port of the C library function. The filename format can also contain the string `[Todo]` which will be selected in the prompt. The default format is `[Todo]-Note-%Y%m%d-%H%M%S.md` to be consistent with previous versions.

## 1.4.16 - 2020-10-01

- Added option to [remove priority indicator from completed tasks](https://github.com/davraamides/todotxt-mode/issues/20)
    The spec suggests this to allow sorting completed tasks by date. I made an option
    in the extension settings so as not to change behavior in the past. The default
    is false/off. See `todotxtmode.removePriorityFromCompletedTasks`.

- Added feature to [hover project file contents and link to file](https://github.com/davraamides/todotxt-mode/issues/17)
    This behaves like the `note:xxx` tag field but for `+project` tags. If there is a file
    named `project.md` for the corresponding `+project` field, the contents will be displayed when hovering over the project field and a link to open it is display in the hover box.

## 1.4.15 - 2020-05-17

- Fix [todotxt-mode.toggleCompletion should not overwrite creation date with completion date](https://github.com/davraamides/todotxt-mode/issues/15)
    Correct handling of creation date should be working now. There is an ambiguity in the [spec](https://github.com/todotxt/todo.txt) where it says "Optional - Creation Date (must be specified if completion date is)" which doesn't really make sense since an incomplete task would never have a completion date. In any case, there is one ambiguity I don't handle: if you have a task with a creation date such as:

        2020-05-17 a new task created today

    and then you "complete" it by just inserting  `"x "` at the beginning of the line, there is no way to tell if that is a completion date or a creation date. Since this extension always adds completion dates when using the `toggleCompletion` command, I assume in this case this date is also a completion date.

    Also added these related features:

    - Highlighting of creation dates controlled by style `todotxtmode.creationDateStyle.light.color` and `todotxtmode.creationDateStyle.dark.color`.
    - Sort tasks by creation date via command `todotxt-mode: Sort by CreationDate`.

## 1.4.14 - 2020-01-25

- Fix [Unable to figure out functionality of moveTasksToProject](https://github.com/davraamides/todotxt-mode/issues/7).
    This feature was never fully implemented but is now complete. The behavior is as follows:
    - If the current task is in a `todo` task file, the task will be removed from the current file and inserted into the appropriate project file.
    - If the task has a project tag and that tag is an *exact* match for a project file (excluding the       leading `+` character and adding in the `.md` extension), the task will be moved without prompting       the user for a file. Case is sensitive so a tag of `+Foo` will match the file `Foo.md` but not `foo.md`.
    - If the task does not have a project or the project tag is not an exact match with a filename, the user will be prompted to select a file.
    - A side effect of this process is that the project file will be opened in VS Code. The project tab become active for a second because the VS Code API does not currently allow loading a document without making it active.
    - The changes to both files are *not* saved. Thus if you make a mistake, you can easily undo both edits.
    - Multiple tasks are not supported right now as prompting for a file in a loop is not ideal, but it's also not clear if assuming all the tasks should be moved to the same project file is the right behavior.

## 1.4.13 - 2020-01-17


src/defaults.ts
export namespace Defaults {

    // The ontext and project patterns are prefixed with non-word boundary (\B) as they
    // begin with non-word chars (+@). The tag pattern is prefixed with a word bounday (\b)
    // as tags begin with a word char.
    export const CONTEXT_REGEX = /\B@[^+@\s]+/g;
    export const PRIORITY_REGEX = /[(][A-Z][)]/g;
    export const PROJECT_REGEX = /\B\+[^+@\s]+/g;
    export const TAG_REGEX = /\b\w+:\w+/g;
    export const COMPLETED_REGEX = /^x .*$/g;

    // put them in a map so sorting by any field is consistent
    export const FIELD_REGEX_MAP = {
        'context': CONTEXT_REGEX,
        'priority': PRIORITY_REGEX,
        'project': PROJECT_REGEX,
        'tag': TAG_REGEX,
    }

    export const CONTEXT_STYLE = {
        light: {
            color: 'rgb(40, 161, 86)'
        },
        dark: {
            color: 'rgb(40, 161, 86)'
        }
    };
    export const PRIORITY_STYLE = {
        light: {
            color: 'rgb(230, 216, 25)'
        },
        dark: {
            color: 'rgb(230, 216, 25)'
        }
    };
    export const PROJECT_STYLE = {
        light: {
            color: 'rgb(25, 172, 230)'
        },
        dark: {
            color: 'rgb(25, 172, 230)'
        }
    };
    export const TAG_STYLE = {
        light: {
            color: 'rgb(179, 58, 172)'
        },
        dark: {
            color: 'rgb(179, 58, 172)'
        }
    };
    export const COMPLETED_STYLE = {
        textDecoration: "line-through",
        opacity: "0.5"
    };
};
